Simple reproducibility with minimum boilerplate train_cli

mauvilsa opened the pull request:
## What does this PR do?

This is a proposal which is much easier to explain with actual code in a pull request. The exact idea is not discussed in a single issue, instead it relates to multiple. Though the most relevant discussion would be the one in #3720.

I have added to the pull request an rst file which describes the proposal in more detail, so please have a look: [lightning_cli.rst](https://github.com/mauvilsa/pytorch-lightning/blob/config_files_argparse/docs/source/lightning_cli.rst).

The main motivation I have for this is to standardize and remove as much as possible boilerplate code in order to have trainings that can be trivially reproduced with a single config file. This config file is automatically saved in the log dir.

The short summary is, users would run a training like:

```bash
    python trainer.py --config config.yaml
```

And then they would be able to reproduce this by running:

```bash
    python trainer.py --config lightning_logs/version_X/config.yaml
```

The point is not rerunning a training that has already been done. But that the config file includes all of the information needed to reproduce it. This is possible because the config file also includes all of the settings from user defined classes.

This proposal overlaps with other already existing features, namely argument parsing and omegaconf. Though I believe it goes way beyond those in a much simpler way, providing several additional advantages. Specifically:

- Single standardized config file including settings for trainer and user defined classes.
- Encourages users to write better code with good type hints and docstrings.
- Implementation of training command line tools require minimum boilerplate code.
- Supports more complex trainer arguments: list, dict, see [test_lightning_cli.py#L157-L165](https://github.com/mauvilsa/pytorch-lightning/blob/config_files_argparse/tests/trainer/test_lightning_cli.py#L157-L165).
- The training tool help automatically includes defaults and descriptions in docstrings.
- Supports parsing settings from command line, config file and environment variables.
- Supports tab argument completion if argcomplete is installed.
- Very small amount code in comparison to the number features it provides.

The unit tests [test_lightning_cli.py](https://github.com/mauvilsa/pytorch-lightning/blob/config_files_argparse/tests/trainer/test_lightning_cli.py) are in part a copy from the existing [test_trainer_cli.py](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/tests/trainer/test_trainer_cli.py) tests. Almost all functionalities from previous code are supported. The only exception is that command line arguments for booleans are more strict, not allowing for example "0" or "t", see selected tests in [test_trainer_cli.py#L110-L119](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/tests/trainer/test_trainer_cli.py#L110-L119). Certainly the support for these could be added, though being more strict could actually be a good thing. With argument completion this is even less of an issue.

Since parts of this proposal are new, I surely hope to get lots of feedback to improve the idea further.

Fixes #3076
Fixes #3720
Fixes #4119

## Before submitting
- [X] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [X] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [X] Did you make sure your PR does only one thing, instead of bundling different changes together? Otherwise, we ask you to create a separate PR for every change.
- [X] Did you make sure to update the documentation with your changes?
- [X] Did you write any new necessary tests? 
- [X] Did you verify new and existing tests pass locally with your changes?
- [x] If you made a notable change (that affects users), did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)?

## PR review
 - [X] Is this pull request ready for review? (if not, please submit in draft mode)

## Next steps:

- Deprecate the current argparse support
- Migrate examples

pep8speaks commented:
Hello @mauvilsa! Thanks for opening this PR. 

* In the file [`pytorch_lightning/utilities/jsonargparse_utils.py`](https://github.com/PyTorchLightning/pytorch-lightning/blob/b04b51593158c1db6658d0a7c00f0b4dde2c93f9/pytorch_lightning/utilities/jsonargparse_utils.py):

> [Line 27:13](https://github.com/PyTorchLightning/pytorch-lightning/blob/b04b51593158c1db6658d0a7c00f0b4dde2c93f9/pytorch_lightning/utilities/jsonargparse_utils.py#L27): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent
> [Line 28:13](https://github.com/PyTorchLightning/pytorch-lightning/blob/b04b51593158c1db6658d0a7c00f0b4dde2c93f9/pytorch_lightning/utilities/jsonargparse_utils.py#L28): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent
> [Line 31:5](https://github.com/PyTorchLightning/pytorch-lightning/blob/b04b51593158c1db6658d0a7c00f0b4dde2c93f9/pytorch_lightning/utilities/jsonargparse_utils.py#L31): [E303](https://duckduckgo.com/?q=pep8%20E303) too many blank lines (2)
> [Line 47:5](https://github.com/PyTorchLightning/pytorch-lightning/blob/b04b51593158c1db6658d0a7c00f0b4dde2c93f9/pytorch_lightning/utilities/jsonargparse_utils.py#L47): [E303](https://duckduckgo.com/?q=pep8%20E303) too many blank lines (2)
> [Line 63:5](https://github.com/PyTorchLightning/pytorch-lightning/blob/b04b51593158c1db6658d0a7c00f0b4dde2c93f9/pytorch_lightning/utilities/jsonargparse_utils.py#L63): [E303](https://duckduckgo.com/?q=pep8%20E303) too many blank lines (2)
> [Line 86:5](https://github.com/PyTorchLightning/pytorch-lightning/blob/b04b51593158c1db6658d0a7c00f0b4dde2c93f9/pytorch_lightning/utilities/jsonargparse_utils.py#L86): [E303](https://duckduckgo.com/?q=pep8%20E303) too many blank lines (2)

* In the file [`tests/trainer/test_trainer_cli_new.py`](https://github.com/PyTorchLightning/pytorch-lightning/blob/b04b51593158c1db6658d0a7c00f0b4dde2c93f9/tests/trainer/test_trainer_cli_new.py):

> [Line 98:5](https://github.com/PyTorchLightning/pytorch-lightning/blob/b04b51593158c1db6658d0a7c00f0b4dde2c93f9/tests/trainer/test_trainer_cli_new.py#L98): [E265](https://duckduckgo.com/?q=pep8%20E265) block comment should start with '# '
> [Line 100:5](https://github.com/PyTorchLightning/pytorch-lightning/blob/b04b51593158c1db6658d0a7c00f0b4dde2c93f9/tests/trainer/test_trainer_cli_new.py#L100): [E265](https://duckduckgo.com/?q=pep8%20E265) block comment should start with '# '
> [Line 102:5](https://github.com/PyTorchLightning/pytorch-lightning/blob/b04b51593158c1db6658d0a7c00f0b4dde2c93f9/tests/trainer/test_trainer_cli_new.py#L102): [E265](https://duckduckgo.com/?q=pep8%20E265) block comment should start with '# '
> [Line 104:5](https://github.com/PyTorchLightning/pytorch-lightning/blob/b04b51593158c1db6658d0a7c00f0b4dde2c93f9/tests/trainer/test_trainer_cli_new.py#L104): [E265](https://duckduckgo.com/?q=pep8%20E265) block comment should start with '# '

Do see the [Hitchhiker's guide to code style](https://goo.gl/hqbW4r)

mauvilsa commented:
I ping a few more people that could be interested in this. @carmocca @odedbd

carmocca commented on docs/source/trainer_cli.rst:
```diff
@@ -0,0 +1,156 @@
+.. code-block:: bash
+
+    # Dump default configuration to have as reference
+    python trainer.py --print-config > default_config.yaml
```
Lightning uses underscore to split argument names (e.g. `--print_config`). I see this argument is defined inside jsonargparse. Can you add an alias for consistency?

mauvilsa replied:
Will do

mauvilsa replied:
Done

Borda commented on docs/source/trainer_cli.rst:
```diff
@@ -0,0 +1,156 @@
+.. code-block:: python
+
+    from pytorch_lightning.utilities.jsonargparse_utils import trainer_cli
+    from mycode import LitModel
```
can we user model from examples or boring model

Borda commented on docs/source/trainer_cli.rst:
```diff
@@ -0,0 +1,156 @@
+required :code:`*_dataloader` methods, a :code:`trainer.py` tool can be as
+simple as:
+
+.. code-block:: python
```
this code with one of PL included model shall run fine, right? consider using `.. testcode::`

mauvilsa replied:
Yes I was considering using `.. testcode::` but preferred to open the pull request to get first some initial feedback. Will do.

mauvilsa replied:
Converted the code snippets to `.. testcode::`. It is the first time I do this, so please let me know if I did something wrong.

mauvilsa replied:
I finally got all `testcode` working.

Borda commented on docs/source/trainer_cli.rst:
```diff
@@ -0,0 +1,156 @@
+    # Create config including only options to modify
+    nano config.yaml
+    # Run training using created configuration
+    python trainer.py --cfg config.yaml
```
with `--` let's use longer more explanatory name like `--config` or `--config_path`

mauvilsa replied:
I agree, better to have a longer name. Will change it.

mauvilsa replied:
Changed it to `--config`. It is not `--config_path` because it is not necessarily a path. It could be directly a string `--config '{"op1": val1, "op2": val2}'`. Or it could even be configured so that it can also be a URL and the config be fetched remotelly, see https://omni-us.github.io/jsonargparse/#parsing-urls.

carmocca commented on docs/source/trainer_cli.rst:
```diff
@@ -0,0 +1,156 @@
+Python ArgumentParser that makes it very simple to implement configurable
+training tools with the same features as :func:`trainer_cli`.
+
+An example of a more complex training tool could be one in which there are
+several independent modules that require configuration. The code for such a case
+could look something like:
```
Can you add an example of how the `--help` of the following snippet would look like?. Is each module inside their own ArgumentGroup by default?

mauvilsa replied:
Yes, each module would have its own group. I will add how the help looks like.

mauvilsa replied:
Added help snippet.

Borda commented on tests/trainer/test_trainer_cli_new.py:
```diff
@@ -0,0 +1,245 @@
+
+
+@pytest.mark.parametrize(['cli_args', 'expected'], [
+    #pytest.param('--auto_lr_find --auto_scale_batch_size power',
+    #             {'auto_lr_find': True, 'auto_scale_batch_size': 'power'}),
+    #pytest.param('--auto_lr_find any_string --auto_scale_batch_size',
+    #             {'auto_lr_find': 'any_string', 'auto_scale_batch_size': True}),
+    #pytest.param('--auto_lr_find t --auto_scale_batch_size ON',
+    #             {'auto_lr_find': True, 'auto_scale_batch_size': True}),
+    #pytest.param('--auto_lr_find 0 --auto_scale_batch_size n',
+    #             {'auto_lr_find': False, 'auto_scale_batch_size': False}),
```
these do not work?

mauvilsa replied:
This is explained in the description. They don't work, but maybe they shouldn't. These are booleans, so their values should be true or false.

mauvilsa replied:
Removed the commented params and added a few more since according to the yaml standard `On`, `Off`, `Yes` and `No` also work.

carmocca commented on docs/source/trainer_cli.rst:
```diff
@@ -0,0 +1,156 @@
+    # Define parser
+    parser = LightningArgumentParser(description='pytorch-lightning trainer',
+                                     parse_as_dict=True)
+    parser.add_trainer_args()
```
What would be used for a basic Python class? `add_class_arguments`? Mind linking its docs somewhere?

mauvilsa replied:
Yes, for a basic class `add_class_arguments` would be used. Will link the relevant documentation.

mauvilsa replied:
Added link.

Borda commented on pytorch_lightning/utilities/jsonargparse_utils.py:
```diff
@@ -0,0 +1,150 @@
+def trainer_cli(
+    model_class: Type[LightningModule],
+    datamodule_class: Type[LightningDataModule] = None,
+    save_config_callback: Type[Callback] = SaveConfigCallback,
```
well if there will be some derived Saver which would take also init arguments it won't work, consider passing instances instead...

mauvilsa replied:
The SaveConfigCallback class requires to be instantiated with the parser and parsed config which are not yet available outside of trainer_cli. It could be done differently, but I guess it would require a more tight integration. Would need to get some ideas about this and what would be the benefits.

Borda replied:
so it means that any custom Saver can't have any init arguments? some each change has to be made with extra child?

mauvilsa replied:
The idea I had here was that people would extend it if for example additionally to saving it in the log directory they would want to log it as an artifact. For this there is no need for new init arguments. What would be an example in which more init arguments are required? Anyway if someone wants something different they could add a custom callback like how @SeanNaren proposes in https://github.com/PyTorchLightning/pytorch-lightning/pull/4492#issuecomment-721194868

carmocca commented on pytorch_lightning/utilities/jsonargparse_utils.py:
```diff
@@ -0,0 +1,150 @@
+            help='Path to a configuration file in json or yaml format.')
+
+
+    def add_trainer_args(
```
The add_*_args methods should include a `blocklist` argument, a list of arguments to exclude. For example, I may want to hide `tpu_cores` from the CLI arguments if my code does not support TPUs

mauvilsa replied:
I was thinking of this, but hadn't implemented it yet. But will do.

mauvilsa replied:
jsonargparse now has a the option to skip some arguments. Though I haven't added yet to the trainer cli.

mauvilsa replied:
I am unsure about this. The config options from the lightning trainer could be considered like a standard, so removing might not be a good idea. In your example, TPUs not supported, instead of hiding wouldn't it be better to extend Trainer, having `tpu_cores` with a docstring saying that it is not supported and if given a value fail with a not supported message?

carmocca replied:
If I understood correctly what you are proposing, it would require duplicating the entire `Trainer` docstring for the `ConstrainedTrainer` subclass with only a few lines of it changed. That seems very cumbersome, especially if you want to have different `ConstrainedTrainer`s for different tasks/environments.

Currently, I am using the following hack to limit the arguments exposed:

https://github.com/carmocca/PyLaia/blob/c99ff97432892c97c43c16aaba253895682714ce/laia/common/arguments.py#L253-L286

mauvilsa replied:
For a case like PyLaia there is already the `skip` option in [add_class_arguments](https://omni-us.github.io/jsonargparse/#jsonargparse.signatures.SignatureArguments.add_class_arguments). But I don't think PyLaia is a good example for this discussion since it is not designed to use the `LightningCLI` in this pull request. Maybe could you explain why you are skipping those 8 arguments to understand if in general it would also make sense to skip them when using `LightningCLI`.

carmocca replied:
Taking your snippet: 

```python
from mymodels import MyModel
from pytorch_lightning.utilities.jsonargparse_utils import LightningCLI
LightningCLI(MyModel)
```
Imagine `MyModel` fails to work with AMP due to some unrelated PyTorch bug. To avoid people getting confused by whatever error is thrown, Im thinking about being able to do

```python
LightningCLI(MyModel, skip=["precision"])
# or maybe using the key prefix to avoid conflicts if we support skipping non-trainer args
LightningCLI(MyModel, skip=["trainer.precision"])
```

If i'm not mistaken, currently to achieve this, either you don't use `TrainerCLI` or you do the following:

```python
class MyArgumentParser(LightningArgumentParser):
    def add_trainer_args(...):
        self.add_class_arguments(trainer_class, skip=["precision"])

class MyLightningCLI(LightningCLI):
    def init_parser(...):
        self.parser = MyArgumentParser(...)

MyLightningCLI(MyModel)
```

mauvilsa replied:
Adding skip is simple. Still my doubt is that by removing arguments from the trainer, the user does not know why this argument is missing. It is supposed to be a lightning trainer so you kind of expect them to be there. Instead of simply removing them it could be something like `LightningCLI(MyModel, unsupported={"precision": "unsupported due to bug in XXX"})`. The option wouldn't be removed, but shown with this message in the help or if attempted to be used.

Maybe I am just overthinking this.

carmocca replied:
supporting `skip` and `unsupported` does not seem exclusive to me. Some may want to just hide it and others to be able to overwrite it.

To me, most CLI users do not know about the `Trainer` class under the hood and won't notice some arguments are missing.

> Maybe I am just overthinking this.

Maybe :smile:

mauvilsa replied:
@carmocca is this a requirement or can we leave this as a future improvement?

carmocca replied:
We can leave it for the future. I would rather not block the PR over this

Borda commented on requirements.txt:
```diff
@@ -5,6 +5,7 @@ torch>=1.3,<1.8
 future>=0.17.1  # required for builtins in setup.py
 # pyyaml>=3.13
 PyYAML>=5.1  # OmegaConf requirement >=5.1
+jsonargparse[signatures]>=3.0.0.dev4  # New trainer_cli requirement
```
using `dev4` does not look very stable...

mauvilsa replied:
Version 3.0.0 will be released very soon. It is still in dev mostly waiting for the feedback from this pull request. It is also already highly tested, see https://codecov.io/gh/omni-us/jsonargparse/src/master/jsonargparse/signatures.py

Borda replied:
can we use also version 2?

mauvilsa replied:
No, version 2 does not include the main feature used for this. Still, if people like this PR we don't need to rush it. We could first try it out with different models and use cases and extending jsonargparse where needed. Only after it is stable and highly tested we include this in lightning.

mauvilsa replied:
The version is now `3.0.0rc1`. Still a bit more of work remaining for the official `3.0.0` release.

mauvilsa replied:
Version 3 has now been released.

Borda commented on pytorch_lightning/utilities/jsonargparse_utils.py:
```diff
@@ -0,0 +1,150 @@
+
+import os
+from typing import Type
+from jsonargparse import ArgumentParser, ActionConfigFile
```
what is the advantage of using this arg parser compare to the native one?

mauvilsa replied:
jsonargparse is the one that provides all the features (config files, arguments from type hints, etc.). The native one does not have any of this.

Borda commented on pytorch_lightning/utilities/jsonargparse_utils.py:
```diff
@@ -0,0 +1,150 @@
+            help='Path to a configuration file in json or yaml format.')
+
+
+    def add_trainer_args(
```
do we need to add each separately, or just pass all three at once?

carmocca replied:
I personally like them separate. But we could include one more method which calls all three.

mauvilsa replied:
The idea of having separate is to give users more control for more advanced use cases, see https://github.com/mauvilsa/pytorch-lightning/blob/config_files_argparse/docs/source/trainer_cli.rst#lightningargumentparser

Borda commented on pytorch_lightning/utilities/jsonargparse_utils.py:
```diff
@@ -0,0 +1,150 @@
+            nested_key: Name of the nested namespace where parsed arguments are stored.
+        """
+        assert issubclass(trainer_class, Trainer)
+        self.add_class_arguments(trainer_class, nested_key)
```
this mimic out gets default args, right?
does it support also argument types and docs as help?

mauvilsa replied:
Yes, it gets the arguments with based on its type hints and the description from the docstring goes to the help. The arguments can have complex types and are validated as such, e.g. `Union[int, Dict[int, int]`.

Borda commented:
@nateraw min have a look, this seems like complete redone our CLI... :]

carmocca commented:
Would we deprecate the current argparse functionality (and bolts implementation) if this lands?

SeanNaren commented:
This is so so epic! Thanks for an awesome PR 👍 

How would this work with shared parameters, i.e I define two models with encoder_layers as an input arg, and I want the same arg to go to both of them? Also curious about inheritance. 

I love the trainer_cli idea, but it seems as soon as you go towards custom args, we lose all the niceties of no boilerplate with the LightningArgsParser. It would be nice to have some form of intermediate that allows us to keep things self contained in PL without having to define the args object/instantiate everything ourselves like normal PL!

mauvilsa commented:
> How would this work with shared parameters, i.e I define two models with encoder_layers as an input arg, and I want the same arg to go to both of them?

I would say there are many ways in which shared parameters could be handled, and the best way to do it maybe would depend on the specific case. This would require a bit more thought. Though a first idea would be that there is a main 

> Also curious about inheritance.

Arguments of base classes are also added correctly, see for instance the example in the unit test https://github.com/omni-us/jsonargparse/blob/master/jsonargparse_tests/signatures_tests.py#L16-L71 . The only requirement is that `*args` and/or `**kwargs` must be present. Not sure if there would be cases which this does not cover, but could be improved further.

> I love the trainer_cli idea, but it seems as soon as you go towards custom args, we lose all the niceties of no boilerplate with the LightningArgsParser. It would be nice to have some form of intermediate that allows us to keep things self contained in PL without having to define the args object/instantiate everything ourselves like normal PL!

@SeanNaren I am not sure what you mean by "allows us to keep things self contained in PL without having to define the args object/instantiate everything ourselves like normal PL". Could you please explain a bit more.

SeanNaren commented:
If I want to specify custom args, I shouldn't need to define all the boilerplate again. I would love to keep it like this (not a really well thought out proposal, but hopefully adds to discussion):

```python
from pytorch_lightning.utilities.jsonargparse_utils import trainer_cli
from mycode import LitModel, LitDataModule

custom_callback = CustomCallback()

trainer_cli(LitModel, LitDataModule, callbacks=[custom_callback])
```

```python
from pytorch_lightning.utilities.jsonargparse_utils import LightningArgumentParser, trainer_cli
from mycode import LitModel, LitDataModule

# Define parser
parser = LightningArgumentParser(description='pytorch-lightning trainer',
                                 parse_as_dict=True)
parser.add_argument('--notification_email', default='will@email.com')

trainer_cli(LitModel, LitDataModule, parser=parser)
```

teddykoker commented:
I think @SeanNaren has a good point. Should we implement a system like this, we would definitely require any arbitrary additional arguments to be added.

Additionally, I think we need to draw the line somewhere where lighting ends, and users can use their own solutions. Right now there are so many solutions for python configuration and CLI: [typer](https://typer.tiangolo.com/), [hydra](https://hydra.cc/), [click](https://click.palletsprojects.com/), I am not sure if it makes sense to build something like this into Lightning, when each user will have their own usecase, and this would add another dependency to worry about breaking.

That being said, I could be wrong, and nevertheless this PR should promote a thoughtful discussion.

mauvilsa commented:
@teddykoker I was certainly expecting comments related to adding a dependency and this probably being the reason why in the end this is not merged. Though putting this aside for a moment, I think the most important topic to discuss here is if lightning should out of the box have this kind of easy reproducibility feature with no boilerplate. People that just want to run some experiment shouldn't need to know about the variety of python solutions for configuration or be required to learn how to use one of them and add add this logic to their code. I personally think the core idea of this PR fits perfectly the lightning philosophy. But this is just one opinion. I would like to hear the opinion about this from as many people as possible.

mauvilsa commented:
If in the end there is a consensus that this is a desired feature, the implementation details would be the second topic to discuss. Many options are possible, including implementing everything in lightning for the sake of avoiding a dependency. Though I would advice against this. Modularity and separation of concerns are basic software engineering principles. Makes more sense to use a package specifically designed for this than reimplementing. I can continue working on this pull request based on the ongoing discussion, though I would do so using jsonargparse, since I am the main developer and I can easily steer its development to whatever the lightning community requires.

mauvilsa commented:
I have committed a relatively big change. The trainer cli is now a class whose methods can be easily extended to customize the cli avoiding boilerplate. Like this it also fits better in lightning since this is how other classes work. This big change relates to the comment from @SeanNaren, so thank you for that. It was spot on. Also thank you @carmocca for all the testing and helping in identifying bugs and expected features in jsonargparse.

Borda commented on docs/source/trainer_cli.rst:
```diff
@@ -0,0 +1,275 @@
+
+.. testcode::
+
+    from pytorch_lightning.utilities.trainer_cli import TrainerCli
```
```suggestion
    from pytorch_lightning.utilities.trainer_cli import TrainerCLI
```
the CLI is already an abbrevation

Borda commented on docs/source/trainer_cli.rst:
```diff
@@ -0,0 +1,275 @@
+
+    pytorch-lightning trainer command line tool
+
+    optional arguments:
+      -h, --help            show this help message and exit
+      --print_config        print configuration and exit
+      --config CONFIG       Path to a configuration file in json or yaml format.
+                            (default: null)
+
+    Customize every aspect of training via flags:
+      ...
+      --trainer.max_epochs MAX_EPOCHS
+                            Stop training once this number of epochs is reached.
+                            (type: int, default: 1000)
+      --trainer.min_epochs MIN_EPOCHS
+                            Force training for at least these many epochs (type: int,
+                            default: 1)
+      ...
+
+    Example encoder-decoder model:
+      --model.encoder_layers ENCODER_LAYERS
+                            Number of layers for the encoder (type: int, default: 12)
+      --model.decoder_layers DECODER_LAYERS
+                            Number of layers for each decoder block (type: List[int],
+                            default: [2, 4])
```
not sure if this keeps here as it means we have to keep it updated and as it not linked to the API with test it would be easy to forget...

carmocca replied:
I proposed adding a snippet of the help. I think it helps users to understand what jsonargparse provides.

I don't think the attributes {max,min}_epochs will change much, it is worth having it to me personally.

mauvilsa replied:
Is there any way we could generate this content dynamically?

carmocca replied:
we only include it as a quick and tiny view of what's provided. I don't think we need to go through that effort

Borda commented on pytorch_lightning/utilities/trainer_cli.py:
```diff
@@ -0,0 +1,221 @@
+            outstream.write(self.config_dump)
+
+
+class TrainerCli:
```
```suggestion
class LightningCLI:
```
the trainer in the name is a bit confusing as it takes Model and datamodule

mauvilsa replied:
I like it. Just changed it.

Borda commented on requirements.txt:
```diff
@@ -5,6 +5,7 @@ torch>=1.3
 future>=0.17.1  # required for builtins in setup.py
 # pyyaml>=3.13
 PyYAML>=5.1  # OmegaConf requirement >=5.1
+jsonargparse[signatures]>=3.0.0rc1  # New trainer_cli requirement
```
I would personally wait with this merge just after regular 3.0.0 it out...

Borda reviewed (commented):
Start to like it :]

williamFalcon commented:
This looks awesome! please hold this until we have a bit of time to discuss whether this is the path we want to take!

carmocca commented on pytorch_lightning/utilities/trainer_cli.py:
```diff
@@ -0,0 +1,221 @@
+
+
+class LightningArgumentParser(ArgumentParser):
+    """Extension of jsonargparse's ArgumentParser for pythorch-lightning"""
```
```suggestion
    """Extension of jsonargparse's ArgumentParser for pytorch-lightning"""
```

mauvilsa replied:
Done

odedbd commented on pytorch_lightning/utilities/trainer_cli.py:
```diff
@@ -0,0 +1,223 @@
+        """Parses command line arguments and stores it in self.config"""
+        self.config = self.parser.parse_args()
+
+    def instantiate_classes(self):
```
I have a situation similar to [this](https://pytorch-lightning.readthedocs.io/en/latest/datamodules.html#using-a-datamodule), where the model has args that are dependent on the DataModule. Not only that, but the DataModule needs to run the prepare_data and setup methods to have the needed parameters ready. 

I guess that can be addressed by inheriting and overriding the instantiate_classes method?

mauvilsa replied:
Yes, this could be addressed by overriding instantiate_classes. But maybe we should split instantiate_classes into three methods, so that only the relevant part needs to be modified.

odedbd replied:
Splitting can be a good solution, but it's also important to instantiate in the right order. I'm not sure if the DM, Model, Trainer order is always the right one, but it is the one I had occasion to use.

mauvilsa replied:
There are now individual `instantiate_{model,datamodule,trainer}` methods. `instantiate_classes` is still there which could be overridden to change order or do things in between.

odedbd replied:
I see intantiate_{model,trainer}, but datamodule seems to be part of prepare_fit_kwargs.

In fact, reading the latest code made me realize that perhaps the cleanest solution would be to move the model instantiation into perpare_fit_kwargs (same as datamodule), and have the user overwrite prepare_fit_kwargs as needed. 

I think that the distinction between initializing the trainer and preparing its kwargs (initializing datamodule/model) is a logical one. Handling any special logic of the datamodule/model interdependency can be encapsulated by asking the user to provide the trainer kwargs. 

Does that make sense to you at all?

mauvilsa replied:
I changed things a bit to be more consistent. There is still prepare_fit_kwargs but now there is an independent instantiate_datamodule method. Right now the methods are rather small which I did so that if someone wants to extend some method, then the change can be kept small. If it looks too fine grained, certainly parts could be moved to perpare_fit_kwargs. Though note that instantiate methods have changed a bit since I implemented a new mode so that a single cli can support multiple models and datamodules.

odedbd commented on pytorch_lightning/utilities/trainer_cli.py:
```diff
@@ -0,0 +1,223 @@
+        self.add_arguments_to_parser(self.parser)
+        self.add_core_arguments_to_parser()
+        self.parse_arguments()
+        self.instantiate_classes()
```
Adding a hook after parsing arguments and before instantiation can help with situations where there are callbacks which are only added to the trainer as needed, based on command line arguments. For example, LearningRateMonitor and EarlyStopping. These can also be addressed by overriding instantiate_classes, but I feel that a hook for processing that parsed arguments is cleaner for that. 

It could even be a dedicated hook, e.g. add_callbacks which gets the callbacks list and config as arguments and returns an updated list. Or it could be similar to before/after init, which are more open ended.

mauvilsa replied:
Maybe the list of callbacks should be given to LightningCLI. This way it is possible to also make the callbacks configurable with the same config file, instead of hardcoding things in the code. So for example in the config file we could have:
```
callbacks:
  LearningRateMonitor:
    log_momentum: true
  EarlyStopping:
    patience: 5
trainer:
  ...
```

odedbd replied:
That's an interesting idea. But doing it like that would remove the dependency between the args and the callbacks. For example, I have no wish to monitor learning rate if the model args specify that the learning rate should be constant.

Also, you won't be able to pass custom callbacks written by the user as part of the config file, right? So some hard-coding can be expected.

I think that being able to specify callbacks or callback params based on args can be important. So, if a way to incorporate it into the process, such as a hook, is possible I'd recommend adding it.

mauvilsa replied:
See comment https://github.com/PyTorchLightning/pytorch-lightning/pull/4492#issuecomment-735705480

mauvilsa replied:
Certainly this does not exclude having a hook for other use cases like the one you mentioned.

mauvilsa replied:
I have added `before_instantiate_classes` method to the cli class.

odedbd reviewed (commented):
First, I'd like to say I like the way this is shaping up. 

I added a few comments regarding more advanced use cases. It could be that for such cases it would make more sense to use the LightningArgumentParser directly, rather than implement overrides or hook for the LightningCLI.

mauvilsa commented on pytorch_lightning/utilities/trainer_cli.py:
```diff
@@ -0,0 +1,223 @@
+        self.init_parser(description, parse_env)
+        self.add_arguments_to_parser(self.parser)
+        self.add_core_arguments_to_parser()
+        self.parse_arguments()
```
True, a hook after add_core_arguments_to_parser could be useful. Though the change of defaults I would personally do different. I would use the `default_config_files` option and have the defaults in a file that is installed with the package. Maybe we should also add `parser_kwargs` that would be used for arbitrary options to instantiate the parser.

mauvilsa replied:
I have added `default_config_files` to the class. Though I haven't added a hook yet.

mauvilsa replied:
I have added `before_parse_arguments` method to the cli class.

mauvilsa replied:
As a side note, in jsonargparse since you have nested namespaces, setting defaults is done with a dictionary, e.g. `parser.set_defaults({'trainer.weights_summary': 'full'})`.

mauvilsa commented:
There is an important improvement with the commit I just made. With previous versions of jsonargparse there were some init args from the `Trainer` class that were not supported. In particular `callbacks` was not supported which is kind of important for the reproducibility. Now all arguments of `Trainer` are supported. Since the `callbacks` is a list (`callbacks: Optional[List[Callback]]`), the way it would be specified in a config file would be for example as follows:

```yaml
trainer:
  callbacks:
    - class_path: pytorch_lightning.callbacks.EarlyStopping
      init_args:
        patience: 5
    - class_path: pytorch_lightning.callbacks.LearningRateMonitor
      init_args:
        ...
```

Essentially any subclass of `Callback` could be provided and configured as long as it can be imported, i.e. defined in a module not generated on the fly.

This is just a feature of jsonargparse which applies to any class used as a type. This means that other arguments would work in a similar way: `logger: Union[LightningLoggerBase, Iterable[LightningLoggerBase], bool]`, `accelerator: Optional[Union[str, Accelerator]]`, `profiler: Optional[Union[BaseProfiler, bool, str]]`. And likewise for user defined classes.

The case of `callbacks` since it is a list, there is one minor downside. If there is a default config file which includes callbacks, if the given config that overrides the defaults includes callbacks, these would totally replace the ones in the defaults. This could be improved so that instead of replacing it does an override merge.

Borda commented:
@mauvilsa apology for the delay, we are quite busy with finishing 1.1 and we would rather have this in 1.2 to have time to let it land safely, but for sure it is a welcome feature!

mauvilsa commented:
Committed a few more improvements.

- Added more methods to LightningCLI and made the whole class more consistent.
- Added a unit test that uses TrialMNISTDataModule and EvalModelTemplate increasing the test coverage to 100%.
- Now using a stable version of jsonargparse.

mauvilsa commented on docs/source/trainer_cli.rst:
```diff
@@ -0,0 +1,275 @@
+
+    from pytorch_lightning.utilities.trainer_cli import TrainerCli
+
+    TrainerCli(MyModel, MyDataModule)
```
Should we just add what it returns, or also show why it would be useful to keep the class in a variable?

mauvilsa replied:
Done

Borda commented:
@mauvilsa just noticed that `jsonargparse` is finally our with stable release, so I think we can move forward...
cc: @PyTorchLightning/core-contributors

mauvilsa commented:
> @mauvilsa just noticed that `jsonargparse` is finally our with stable release, so I think we can move forward...

Yes, jsonargparse is now stable and since lightning v1.1.0 is out, it is a good time to have get a bit more feedback here. Some doubts that I have and not yet finished points are the following:

- Try out LightningCLI with a few varied use cases.
- There are some pending comments from previous reviews that could use more viewpoints.
- Maybe LightningCLI instead of `**kwargs` should receive `trainer_kwargs` and `parser_kwargs` so that some simple cases can be handled without needing to extend class methods.
- lightning_cli.rst is not yet included in any toctree. Where should this be included?
- lightning_cli.rst requires extending, for example:
  * Explain configuration of classes such as Callback.
  * Better explanation of methods in LightningCLI.
  * Video that showcases reduction of boilerplate?

awaelchli commented on requirements.txt:
```diff
@@ -5,6 +5,7 @@ torch>=1.3
 future>=0.17.1  # required for builtins in setup.py
 # pyyaml>=3.13
 PyYAML>=5.1  # OmegaConf requirement >=5.1
+jsonargparse[signatures]>=3.0.1  # LightningCLI requirement
```
Is this required for the Lightning installation or can it be moved to extras?

mauvilsa replied:
Right now the argument parser and the cli class are completely independent from everything else. So the jsonargparse requirement could be moved to an extra without any issue. However, I don't know if you might want to use the new argument parser to replace some of the old parser functionality. Depending on that this requirement could be needed here. It all depends on what the core contributors decide. Based on what you want I could advice on different options.

mauvilsa replied:
Any further comment on this?

carmocca replied:
> I don't know if you might want to use the new argument parser to replace some of the old parser functionality

If this PR lands, we will want to replace the old one. The PRs would be:

Addition (this one) -> Deprecation of current implementation -> Deletion of current implementation

So I think this PR should include the install in extras and the deletion PR should move it to the base requirements.

mauvilsa replied:
I moved the requirement to extras. But now some doctests are failing because jsonargparse is not found. I have no idea how to fix this.

carmocca replied:
@Borda can you help?

Borda replied:
yeah, I ll plan to spend Monday to get this over the finish line :]

odedbd commented:
> There is an important improvement with the commit I just made. With previous versions of jsonargparse there were some init args from the `Trainer` class that were not supported. In particular `callbacks` was not supported which is kind of important for the reproducibility. Now all arguments of `Trainer` are supported. Since the `callbacks` is a list (`callbacks: Optional[List[Callback]]`), the way it would be specified in a config file would be for example as follows:
> 
> ```yaml
> trainer:
>   callbacks:
>     - class_path: pytorch_lightning.callbacks.EarlyStopping
>       init_args:
>         patience: 5
>     - class_path: pytorch_lightning.callbacks.LearningRateMonitor
>       init_args:
>         ...
> ```
> 
> Essentially any subclass of `Callback` could be provided and configured as long as it can be imported, i.e. defined in a module not generated on the fly.
> 
> This is just a feature of jsonargparse which applies to any class used as a type. This means that other arguments would work in a similar way: `logger: Union[LightningLoggerBase, Iterable[LightningLoggerBase], bool]`, `accelerator: Optional[Union[str, Accelerator]]`, `profiler: Optional[Union[BaseProfiler, bool, str]]`. And likewise for user defined classes.
> 
> The case of `callbacks` since it is a list, there is one minor downside. If there is a default config file which includes callbacks, if the given config that overrides the defaults includes callbacks, these would totally replace the ones in the defaults. This could be improved so that instead of replacing it does an override merge.

Just to make sure - this would work for custom callbacks as well, correct? At least as long as their import path is valid at runtime. I definitely like the ability to configure these in a config file by specifying the import path and the parameters.

mauvilsa commented:
> Just to make sure - this would work for custom callbacks as well, correct?

Correct. This should already work with custom callbacks.

mpariente commented:
Any update on this? It looks like a great direction!

mauvilsa commented:
In the last commits additional to improvements related to various comments, worth mentioning are the following:
- In [lightning_cli.rst](https://github.com/mauvilsa/pytorch-lightning/blob/config_files_argparse/docs/source/lightning_cli.rst) it is now explained how specify callbacks in a config file and other arguments that have a class as type.
- Now it is possible to specify a mode for so that a single cli can be used for multiple models and datamodules (subclasses of the given base classes). This is also explained in [lightning_cli.rst](https://github.com/mauvilsa/pytorch-lightning/blob/config_files_argparse/docs/source/lightning_cli.rst). @nathanpainchaud

mpariente commented on docs/source/lightning_cli.rst:
```diff
@@ -0,0 +1,341 @@
+
+.. code-block:: yaml
+
+    model:
+      class_path: mycode.mymodels.MyModel
+      init_args:
+        decoder_layers:
+        - 2
+        - 4
+        encoder_layers: 12
```
To modify these from the command line, would I have to do `--model.init_args.encoder_layers 13`?

mauvilsa replied:
Yes

mpariente replied:
Would it make sense to forward `--model.encoder_layers` to `--model.init_args.encoder_layers` ? IMO, it would.

mauvilsa replied:
I am not sure if there is a simple/clean way to implement this, but I will think about it. Could you please create a feature request for this in jsonargparse.

mauvilsa replied:
I will mark this as resolved since it is not related to the code in pytorch-lightning. This would need to be done in jsonargparse.

mpariente commented on docs/source/lightning_cli.rst:
```diff
@@ -0,0 +1,341 @@
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+In the previous examples :class:`~pytorch_lightning.utilities.cli.LightningCLI`
+works only for a single model and datamodule class. However, there are many
+cases in which the objective is to easily be able to run many experiments for
+multiple models and datasets. For these cases the tool can be configured such
```
> However, there are many cases in which the objective is to easily be able to run many experiments for multiple models and datasets.

I fully agree with that. 

However, the current solution requires the user to create a shared config for each model/data/trainer combination, which doesn't scale. 

Could we have a config file for each part? So that all combinations can be made from the command line? 

So that something like this works out of the box
```bash
python trainer.py --model modelA.yml --data dataA.yml 
python trainer.py --model modelA.yml --data dataB.yml 
python trainer.py --model modelB.yml --data dataA.yml 
python trainer.py --model modelB.yml --data dataB.yml 
```

mauvilsa replied:
What you suggest should already be working. I haven't tested it explicitly with LightningCLI, but that feature is enabled. The relevant part of the code is https://github.com/omni-us/jsonargparse/blob/master/jsonargparse/signatures.py#L301

odedbd commented on pytorch_lightning/utilities/cli.py:
```diff
@@ -0,0 +1,274 @@
+    def instantiate_classes(self):
+        """Instantiates the classes using settings from self.config"""
+        self.config_init = self.parser.instantiate_subclasses(self.config)
+        self.instantiate_model()
```
I still feel it is better to change the order here, so that datamodule would be instantiated before the model. There are common use cases where the model construction depends on the data used (especially for the input/output layers). I haven't yet found a use case where the datamodule depends on the model. While it is true that the use can still override the whole of instantiate classes as needed, I feel that more boilerplate can be saved if the order is datamodule, then model, then trainer.

mauvilsa replied:
Done

odedbd reviewed (approved):
great, thanks!

Borda commented on pytorch_lightning/utilities/cli.py:
```diff
@@ -0,0 +1,274 @@
+        self,
+        model_class: Type[LightningModule],
+        datamodule_class: Type[LightningDataModule] = None,
+        save_config_callback: Type[Callback] = SaveConfigCallback,
```
why it is needed here as type?

mauvilsa replied:
Type is required because the argument must be the class. Not an instance of the class.

mauvilsa replied:
I changed the type to be `Type[SaveConfigCallback]` since it is not any callback that can be given, it must be a callback that is instantiated with `parser` and `config`.

Borda commented on pytorch_lightning/utilities/cli.py:
```diff
@@ -0,0 +1,274 @@
+        self.parser.save(self.config, config_path, skip_none=False)
+
+
+class LightningCLI:
```
it seems that it has many empty shells/methods, so what is the use-case of sub-classing this?

mauvilsa replied:
There could be many use cases. One is the example of sending email notifications. Another, to implement a more complex trainer, multiple stages, multiple datasets, multiple modules for different parts of a big model, etc. Another, to instantiate a model based on details only available after loading the data. ...

Borda reviewed (commented):
mind check some of past comments and mark them as resolved if they are done... :]

mauvilsa commented on docs/source/lightning_cli.rst:
```diff
@@ -0,0 +1,341 @@
+            """
+            pass
+
+    class MyDataModule(LightningDataModule):
+        pass
+
+    def send_email(address, message):
+        pass
```
This is needed so that the testcode with the email address example works.

mauvilsa commented on pytorch_lightning/utilities/cli.py:
```diff
@@ -0,0 +1,274 @@
+    def add_trainer_args(
+        self,
+        trainer_class: Type[Trainer] = Trainer,
+        nested_key: str = 'trainer'
```
For example imagine that I want to create a cli to run a two or more stages training. The settings for each stage should be different. So I would call `add_trainer_args` multiple times, each with a different `nested_key`. This is also one reason why it can make sense to have a method specific for each class.

mauvilsa commented on requirements/extra.txt:
```diff
@@ -8,3 +8,4 @@ onnx>=1.7.0
 onnxruntime>=1.3.0
 hydra-core>=1.0
 https://github.com/PyTorchLightning/fairscale/archive/pl_1.1.0.zip
+jsonargparse[signatures]>=3.3.1  # LightningCLI requirement
```
There were comments saying that we should move this to extras. But since LightningCLI is supposed to be extendable by users it seems important that the api appears in the documentation. If I add a check then when building the docs it wouldn't find the classes so wouldn't be included, no? That is what I understood from the errors in the doc builds, since it seems that to build the docs does not preinstall the extras. Please correct me if I am understanding something wrong.

mauvilsa replied:
I have modified pytorch_lightning.utilities.cli so that it can be imported even when jsonargparse not available.

mauvilsa commented on pytorch_lightning/utilities/cli.py:
```diff
@@ -0,0 +1,274 @@
+            help='Path to a configuration file in json or yaml format.'
+        )
+
+    def add_trainer_args(
```
I have just pushed a change so that there is only one method to add arguments, now called `add_lightning_class_args` so that the purpose is clear in comparison to the methods of the base class. I also added some asserts to `LightningCLI` since by having a single add method people could make a mistake giving a class to the wrong argument so the asserts check this. @Borda see how it looks now.

mauvilsa commented:
@Borda only comments from you are pending. Please have a look.

mauvilsa commented:
@Borda please let me know if I can mark as resolved your two pending questions and if there is anything else to do.

Borda commented on docs/source/common/lightning_cli.rst:
```diff
@@ -0,0 +1,341 @@
+Trainer Callbacks and arguments with class type
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+A very important argument of the
```
we can keep 120 line length

carmocca replied:
I believe he means "change all the rst file so that wrapping is at 120"

carmocca commented on pytorch_lightning/utilities/cli.py:
```diff
@@ -0,0 +1,245 @@
+
+    def instantiate_trainer(self):
+        """Instantiates the trainer using self.config_init['trainer']"""
+        self.trainer_kwargs.update(self.config_init['trainer'])
```
This seems to override any arguments passed to `trainer_kwargs`, thus making it useless as `self.config_init['trainer']` will contain all trainer arguments

mauvilsa replied:
True. Will change it.

mauvilsa replied:
I have thought about possible options. It does not make sense that `trainer_kwargs` given to `LightningCLI` have precedence over what is given in a config file or command line. And it would be very confusing if `trainer_kwargs` would only work for some, e.g. override only if default value is `None`. But the current `trainer_kwargs` is useful to set some `callbacks` so that they always available without need to add them in a config file. So I can think of only two sensible solutions. Remove `trainer_kwargs` or change it to be `trainer_callbacks`. If we remove it, then to add persistent callbacks would require to extend a method. Maybe changing to `trainer_callbacks` might be preferable since I guess this can be quite common. What do you think?

carmocca replied:
> It does not make sense that trainer_kwargs given to LightningCLI have precedence over what is given in a config file or command line.

I agree

But it seems that `trainer_kwargs` gets overriden even if you provide no config/command-line arguments, for example:

```python
from pytorch_lightning.utilities.cli import LightningCLI
from tests.helpers.boring_model import BoringDataModule, BoringModel

cli = LightningCLI(
    BoringModel, BoringDataModule,
    trainer_kwargs={'fast_dev_run': True}
)
```

This will not run `fast_dev_run`

Why don't we write `self.config_init['trainer']` only if there is a config?

mauvilsa replied:
Okay, so from what I am understanding it would work like overriding the defaults of the parser and this would only be overridden if given in a config file or command line. For this to work with the correct behavior it needs to be done a bit differently. Not by modifying a kwargs dict. But might be simple anyway.

carmocca commented on pytorch_lightning/utilities/cli.py:
```diff
@@ -0,0 +1,245 @@
+            env_parse: Whether environment variable parsing is enabled.
+            parser_kwargs: Additional arguments to instantiate LightningArgumentParser.
+            subclass_mode_model: Whether model can be any `subclass
+                <https://omni-us.github.io/jsonargparse/#classes-as-type>`_ of the
```
All the links have been moved and need to be updated

mauvilsa replied:
Fixed

williamFalcon commented:
this looks amazing. 
can we land it in the next few days?

Borda commented:
@mauvilsa Great work here, approaching the finish line ;]

mauvilsa commented:
Added a unit test for callbacks as argument from the command line. Also included in the documentation the need for the all extras require.

awaelchli commented:
Is LightningCLI restricted to Lightning classes?
So `some_object = LightningCLI(MyCustomClass)` won't work, even if type annotations are given? Will it be possible to support profilers, loggers etc too?
Why are the Trainer methods embedded inside the cli? Is it not possible to call `trainer.fit()` or test at the desired place?
I only read the docs so far, so apologies but it sounds like this is not possible.

mauvilsa commented:
> Is LightningCLI restricted to Lightning classes? So `some_object = LightningCLI(MyCustomClass)` won't work, even if type annotations are given?

It works for custom classes. That is the idea. People implement their own modules and data modules and then use the cli. The only requirement is that the classes inherit from `LightningModule` and `LightningDataModule`. This is the requirement for lightning's `Trainer`, right?

> Will it be possible to support profilers, loggers etc too?

Yes, they should be working. There are no unit tests specifically for `profilers` and `loggers`. But these have type hints similar to `callbacks` for which there are unit tests so there is no reason why it wouldn't work. Moreover there are several unit tests in jsonargparse that already check this kind of behavior. Trainer has too many parameters and possible values for me to test everything. It is needed that people try it out in their use cases to identify problems.

> Why are the Trainer methods embedded inside the cli? Is it not possible to call `trainer.fit()` or test at the desired place?

I am not entirely sure what you mean. The purpose of `LightningCLI` is to load configuration and then run `trainer.fit` avoiding to write this boilerplate. If you want to do something slightly different then you can implement a new cli inheriting from `LightningCLI` and only needing to modify some of its methods.

> I only read the docs so far, so apologies but it sounds like this is not possible.

No problem. I am glad that someone else is looking at it. I guess it will become a bit clearer if you try it out with a simple example.

carmocca reviewed (approved):
Ready to go! 💃

tchaton reviewed (approved):
LGTM !

Borda commented:
🎉   💟  🐰
